pyseer/enet_predict.py

```diff
@@ -0,0 +1,183 @@
+
+    return parser.parse_args()
+
+def main():
+
+    options = get_options()
+
+    # Read in model pickle
+    with open(options.model, 'rb') as pickle_obj:
+        model_dict, continuous = pickle.load(pickle_obj)
+    try:
+        intercept = model_dict.pop('intercept')[1]
+    except KeyError as e:
+        sys.stderr.write("Intercept not found in model\n")
+        intercept = 0
+
+    # Read in samples, start building predictions
+    samples = []
+    with open(options.samples, 'r') as sample_file:
+        for sample in sample_file:
+            samples.append(sample.rstrip())
+
+    p = pd.DataFrame(data=np.full(len(samples), intercept),
+                     index=samples,
+                     columns=['prediction'])
+    predictions = np.array(p.values, dtype=np.float)
+
+    # Read in covariates
+    if options.covariates is not None:
+        cov = load_covariates(options.covariates,
+                              options.use_covariates,
+                              p)
+        if cov is None:
+            sys.exit(1)
+        else:
+            for covariate in cov:
+                pred_beta = model_dict.pop(covariate, (0, 0))
+                if pred_beta[1] != 0:
+                    predictions += (cov[covariate] * pred_beta[1]).reshape(-1, 1)
+
+    # Open variant file
+    sample_order = []
+    all_strains = set(p.index)
+    burden_regions = deque([])
+    burden = False
+
+    if options.kmers:
+        var_type = "kmers"
+        var_file = options.kmers
+    elif options.vcf:
+        var_type = "vcf"
+        var_file = options.vcf
+        if options.burden:
+            burden = True
+    else:
+        var_type = "Rtab"
+        var_file = options.pres
+
+    infile, sample_order = open_variant_file(var_type, var_file, options.burden, burden_regions, options.uncompressed)
+
+    # Read in all variants - only keep if matching name
+    sys.stderr.write("Reading variants from input\n")
+    pbar = tqdm(unit="variants")
+    while True:
+        eof, k, var_name, kstrains, nkstrains, af = read_variant(
+                                        infile, p, var_type,
+                                        burden, burden_regions,
+                                        options.uncompressed, all_strains,
+                                        sample_order, keep_list = model_dict.keys())
+
+        # check for EOF
+        if eof:
+            pbar.close()
+            break
+        else:
+            pbar.update(1)
+
+        # return 0 if not found. remove from dict if found
+        pred_beta = model_dict.pop(var_name, (0, 0))
+        if pred_beta[1] != 0:
+            predictions += (k * pred_beta[1]).reshape(-1, 1)
+
+    # Note those variants which did not appear - impute
+    for missing in model_dict.keys():
+        sys.stderr.write("Could not find covariate/variant " + missing + " in input file\n")
+        predictions += model_dict[missing][0] * model_dict[missing][1]
```

Comment:
So if I understood correctly here we are using the AF and the betas to impute missing variants. Why would a variant be missing in the first place? Because the sample file is smaller than the total list of strains? Also, in `__main__.py` the first element might be either AF (line 626) or the average of the covariates (line 632)? Have I got it right?

Reply:
The most likely reason for a variant being missing would be trying to apply a fitted model to perform prediction on a new dataset.
The first element should be the mean of the predictor x i.e. E(x) / <x>. So I think both the AF or average of the covariate would be right